File: apps/picopad/lib/picopad/crc8.ex
defmodule Picopad.CRC8 do
  @polynomial 0x07
  @init 0x00

  def calculate(data) when is_binary(data) do
    data
    |> :binary.bin_to_list()
    |> Enum.reduce(@init, &update/2)
  end

  defp update(byte, crc) do
    crc = Bitwise.bxor(crc, byte)

    Enum.reduce(0..7, crc, fn _, acc ->
      if Bitwise.band(acc, 0x80) != 0 do
        Bitwise.band(Bitwise.bsl(acc, 1) |> Bitwise.bxor(@polynomial), 0xFF)
      else
        Bitwise.band(Bitwise.bsl(acc, 1), 0xFF)
      end
    end)
  end
end

File: firmware/picopad/src/crc8.c
#include "crc8.h"

uint8_t crc8_update(uint8_t crc, uint8_t byte) {
    crc ^= byte;
    for (int i = 0; i < 8; i++) {
        if (crc & 0x80) {
            crc = (crc << 1) ^ CRC8_POLYNOMIAL;
        } else {
            crc <<= 1;
        }
    }
    return crc;
}

uint8_t crc8(const uint8_t *data, size_t length) {
    if (data == NULL) {
        return CRC8_INIT;
    }
    uint8_t crc = CRC8_INIT;
    for (size_t i = 0; i < length; i++) {
        crc = crc8_update(crc, data[i]);
    }
    return crc;
}
